src/lib/string_pool.rs

```diff
@@ -0,0 +1,509 @@
+use std::mem::swap;
+
+pub const INIT_BLOCK_SIZE: usize = init_block_size_const();
+
+#[cfg(feature = "mozilla")]
+const fn init_block_size_const() -> usize {
+    // FIXME: should be `offset_of(BLOCK, s)`, but that's not supported yet,
+    // so we over-estimate its offset
+    1024 - (std::mem::size_of::<BLOCK>() / std::mem::size_of::<XML_Char>())
+}
+
+#[cfg(not(feature = "mozilla"))]
+const fn init_block_size_const() -> usize {
+    1024
+}
+
+fn poolBytesToAllocateFor(mut blockSize: c_int) -> size_t {
+    /* Unprotected math would be:
+     ** return offsetof(BLOCK, s) + blockSize * sizeof(XML_Char);
+     **
+     ** Detect overflow, avoiding _signed_ overflow undefined behavior
+     ** For a + b * c we check b * c in isolation first, so that addition of a
+     ** on top has no chance of making us accept a small non-negative number
+     */
+    let stretch = ::std::mem::size_of::<XML_Char>() as u64; /* can be 4 bytes */
+    if blockSize <= 0 {
+        return 0;
+    }
+    if blockSize > (INT_MAX as c_ulong).wrapping_div(stretch) as c_int {
+        return 0;
+    }
+    let stretchedBlockSize: c_int = blockSize * stretch as c_int;
+    let bytesToAllocate: c_int =
+        (12 as c_ulong).wrapping_add(stretchedBlockSize as c_uint as c_ulong) as c_int;
+    if bytesToAllocate < 0 {
+        return 0;
+    }
+    bytesToAllocate as size_t
+}
+
+rental! {
+    mod rental_pool {
+        use super::*;
+
+        #[rental(debug)]
+        pub(crate) struct InnerStringPool {
+            bump: Box<Bump>,
+            current_bump_vec: RefCell<BumpVec<'bump, XML_Char>>,
+        }
+    }
+}
+
+use rental_pool::InnerStringPool;
+
+/// A StringPool has the purpose of allocating distinct strings and then
+/// handing them off to be referenced either temporarily or for the entire length
+/// of the pool.
+pub(crate) struct StringPool(Option<InnerStringPool>);
+
+impl StringPool {
+    pub(crate) fn try_new() -> Result<Self, ()> {
+        let bump = Bump::try_new().map_err(|_| ())?;
+        let boxed_bump = Box::try_new(bump).map_err(|_| ())?;
+
+        Ok(StringPool(Some(InnerStringPool::new(
+            boxed_bump,
+            |bump| RefCell::new(BumpVec::new_in(&bump)),
+        ))))
+    }
+
+    /// # Safety
+    ///
+    /// The inner type is only ever None in middle of the clear()
+    /// method. Therefore it is safe to use anywhere else.
+    fn inner(&self) -> &InnerStringPool {
+        self.0.as_ref().unwrap_or_else(|| unsafe {
+            std::hint::unreachable_unchecked()
+        })
+    }
+
+    /// Determines whether or not the current BumpVec is empty.
+    pub(crate) fn is_empty(&self) -> bool {
+        self.inner().rent(|vec| vec.borrow().is_empty())
+    }
+
+    /// Determines whether or not the current BumpVec is full.
+    pub(crate) fn is_full(&self) -> bool {
+        self.inner().rent(|vec| self.rented_is_full(&vec.borrow()))
+    }
+
+    fn rented_is_full(&self, vec: &BumpVec<XML_Char>) -> bool {
```

Comment:
It really feels to me like all the `rented_NNN` functions should be methods of a separate rented object. How about doing this:
```Rust
...
        pub(crate) struct InnerStringPool {
            bump: Box<Bump>,
            current_bump_vec: RefCell<RentedBumpVec<'bump>>,
        }
...

struct RentedBumpVec<'bump>(BumpVec<'bump, XML_Char>);

impl<'bump> RentedBumpVec<'bump> {
  // All rented_X go here as methods of RentedBumpVec (without rented_ prefix), e.g.
  fn is_full(&self) -> bool {
    self.0.len() == self.0.capacity()
  }
  // ...all the others, like `append_char`
}
```

Reply:
This is a good idea!